pkg/util/connectionpool/grpc/pool.go

```diff
@@ -0,0 +1,311 @@
+		grpc.WithTransportCredentials(insecure.NewCredentials()),
+		grpc.WithCodec(GetCodecInstance())}
+	_ connectionpool.Pool = (*Pool)(nil)
+)
+
+func NewPool(spec *Spec) *Pool {
+	if spec.InitConnNum == 0 {
+		spec.InitConnNum = defaultInitNum
+	}
+	if spec.BorrowTimeout == 0 {
+		spec.BorrowTimeout = defaultBorrowTimeout
+	}
+	if spec.ConnectTimeout == 0 {
+		spec.ConnectTimeout = defaultConnectTimeout
+	}
+	if spec.MaxConnsPerHost == 0 {
+		spec.MaxConnsPerHost = defaultMaxIdleConnsPerHost
+	}
+	if spec.DialOptions == nil {
+		spec.DialOptions = defaultDialOpts
+	}
+
+	logger.Infof("created grpc connection pool with spec %+v", spec)
+	return &Pool{
+		spec:       spec,
+		connsMutex: &sync.Mutex{},
+		conns:      map[string]*segmentConn{},
+	}
+}
+
+func NewPoolWithFactory(spec *Spec, factory connectionpool.CreateConnFactory) *Pool {
+	if factory == nil {
+		panic("the factory shouldn't be nil")
+	}
+	if spec.InitConnNum == 0 {
+		spec.InitConnNum = defaultInitNum
+	}
+	if spec.BorrowTimeout == 0 {
+		spec.BorrowTimeout = defaultBorrowTimeout
+	}
+	if spec.ConnectTimeout == 0 {
+		spec.ConnectTimeout = defaultConnectTimeout
+	}
+	if spec.MaxConnsPerHost == 0 {
+		spec.MaxConnsPerHost = defaultMaxIdleConnsPerHost
+	}
+	if spec.DialOptions == nil {
+		spec.DialOptions = defaultDialOpts
+	}
+
+	logger.Infof("created grpc connection pool with factory and spec %+v", spec)
+	return &Pool{
+		spec:       spec,
+		connsMutex: &sync.Mutex{},
+		conns:      map[string]*segmentConn{},
+		factory:    factory,
+	}
+}
+
+func (s *segmentConn) buildConnectionWithChannel(factory connectionpool.CreateConnFactory, spec *Spec, errCh chan<- error) {
+	s.Lock()
+	defer s.Unlock()
+	waited := int(atomic.LoadInt32(&s.waitedNum))
+	if waited <= len(s.connCh) || len(s.connCh) == cap(s.connCh) {
+		return
+	}
+	// it means all conn in used, so we pick random conn
+	if len(s.connMap) == int(spec.MaxConnsPerHost) {
+		for c := range s.connMap {
+			if c.GetState() == connectivity.Shutdown {
+				delete(s.connMap, c)
+				continue
+			}
+			s.connCh <- &Connection{ClientConn: c, segment: s, needBack: false}
+			return
+		}
+	}
+
+	for i := 0; i < int(spec.InitConnNum); i++ {
+		rawConn, err := buildConnection(s.addr, factory, spec)
+		if err != nil {
+			errCh <- status.Errorf(codes.Internal, "couldn't create connection for addr %s, cause: %s", s.addr, err.Error())
+			logger.Infof("not create connection for addr %s, cause: %s", s.addr, err.Error())
+			return
+		}
+		c := &Connection{
+			segment:    s,
+			ClientConn: rawConn,
+			needBack:   true,
+		}
+		s.connCh <- c
+		s.connMap[rawConn] = nil
+		// other goroutine return conn
+		if len(s.connMap) >= int(spec.MaxConnsPerHost) {
+			break
+		}
+	}
+}
+
+func (m *Pool) Get(addr string) (interface{}, error) {
+	timeout, cancelFunc := context.WithTimeout(context.Background(), m.spec.BorrowTimeout)
+	defer cancelFunc()
+	segment := m.conns[addr]
+	if segment == nil {
+		m.connsMutex.Lock()
+		if m.conns[addr] == nil {
```

Comment:
the pattern of the `connsMutex` usage is wrong, either we don't need it, or there's a race condition.